news/_posts/2020-11-01-prestashop-in-2019-and-beyond-part-4-connecting-the-dots.md

```diff
@@ -0,0 +1,182 @@
+
+The Core API is the better replacement for the obsolete [legacy class override system][class-override], which will progressively be phased out as legacy classes are removed from the root namespace.
+
+### A domain-oriented API
+
+Consider the following application use cases:
+
+- Create a customer through a form in BO.
+- Load customer information for update.
+- Issue a partial refund in an order.
+- Activate debug mode.
+
+Each one of these is a unique and complete user interaction with the system. Whenever a request is issued to the backend, it can usually be matched with exactly one use case. 
+
+The Core API is built around these use cases. This means that all its operations are  based on the things the software does from a user standpoint, rather than what its _underlying components_ are able to do. This is a subtle, yet very important difference.
+
+Some would argue that Core API operations should be optimized for reuse and composition. This is normal: as developers, we are used to constantly search for abstractions and generalizations so that we can build reusable components. We don't like duplication because it's inefficient and prone to inconsistencies. 
+
+A low-level API would be the obvious answer, because it's optimized for reusability. After all, use cases are built on top of low level components, and the more low level components there are the less code you need to write whenever you add a new use case. It also helps reduce inconsistencies, because each single thing is done in a single way by a single component, and if that process has to change, you only need to change it in a single place, and then it's changed consistently everywhere. But there's a catch: what if you need to make it work _slightly_ differently for one use case only?
+
+A high-level API, on the other hand, is optimized for customizability. This kind of API answers to a single use case, and nothing else. This provides much more freedom for customization compared to the low-level API, because modifications made to a high level API will be less prone to provoke unexpected side effects in other parts of the system. Its downside, however, is the need for more code, and potentially duplicate logic when two use cases are similar (e.g: single delete and bulk delete).
+
+The Core API minimizes this problem by delegating most of the business logic to lower level Domain components. Therefore, a domain-oriented Core API works like a controller or interface to a domain-oriented Core. It doesn't implement logic, it works as a facade to other services which implement the logic themselves.
```

Comment:
I suggest to add code examples of the aforementioned 4 usecases to "close the loop" and show the path "User usecase => building => API"

Reply:
My first reaction was "okay, let's do that", then I realized that it would require a massive amount of code that no one will ever read 😅

Reply:
Maybe we could provide link to a gist ? 🤔